File: 725 Split Linked List in Parts/solution_1.cpp
/*
  tc - O(N + k)
  sc - O(N + k)
*/

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    vector<ListNode*> splitListToParts(ListNode* head, int k) {

        vector<ListNode *> ans(k, nullptr);
        int cnt = 0;
        ListNode *newHead = head;

        while (head)
        {
            cnt++;
            head = head->next;
        }

        int p = cnt / k;
        int rem = cnt % k, idx = 0;
        vector<int> cnts(k, p);

        while (rem--)
        {
            cnts[idx++]++;
        }

        ListNode *ptr = newHead;
        idx = 0;

        while (idx < k && ptr)
        {
            ans[idx] = ptr;
            
            for (int i=0;i<cnts[idx]-1;i++)
            {
                ptr = ptr->next;
            }

            ListNode *next = ptr->next;
            ptr->next = NULL;
            ptr = next;
            idx++;
        }

        return ans;
    }
};

File: 2326 Spiral Matrix IV/solution_1.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public int[][] spiralMatrix(int m, int n, ListNode head) {
        
        int[][] ans = new int[m][n];
        for (int i=0;i<m;i++)
            for (int j=0;j<n;j++)
                ans[i][j] = -1;
                
        int i = 0, j = 0, i_min = 0, i_max = m-1, j_min = 0, j_max = n - 1, dir = 0;

        while (head != null)
        {
            ans[i][j] = head.val;
            head = head.next;

            if (dir == 0) // right
            {
                if (j == j_max)
                {
                    i_min = ++i;
                    dir = 1;
                }
                else
                {
                    j++;
                }
            }
            else if (dir == 1) // down
            {
                if (i == i_max)
                {
                    j_max = --j;
                    dir = 2;
                }
                else
                {
                    i++;
                }
            }
            else if (dir == 2) // left
            {
                if (j == j_min)
                {
                    i_max = --i;
                    dir = 3;
                }
                else
                {
                    j--;
                }
            }
            else // up
            {
                if (i == i_min)
                {
                    j_min = ++j;
                    dir = 0;
                }
                else
                {
                    i--;
                }
            }
        }

        return ans;
    }
}
